contracts/src/token/erc721/extensions/wrapper.rs

```diff
@@ -0,0 +1,887 @@
+            initial_wrapped_balance - U256::from(tokens)
+        );
+
+        assert_eq!(
+            erc721_contract
+                .sender(contract.address())
+                .balance_of(contract.address())
+                .motsu_unwrap(),
+            initial_contract_balance - U256::from(tokens)
+        );
+    }
+
+    #[motsu::test]
+    fn on_erc721_received_reverts_when_invalid_operator(
+        contract: Contract<Erc721WrapperTestExample>,
+        erc721_contract: Contract<Erc721>,
+        alice: Address,
+    ) {
+        let token_id = random_token_ids(1)[0];
+
+        contract.init(alice, |contract| {
+            contract.wrapper.underlying.set(erc721_contract.address());
+        });
+
+        let invalid_operator = alice;
+
+        let err = contract
+            .sender(invalid_operator)
+            .on_erc721_received(
+                invalid_operator,
+                alice,
+                token_id,
+                vec![].into(),
+            )
+            .motsu_expect_err("should return Error::UnsupportedToken");
+
+        assert!(matches!(
+            err,
+            Error::UnsupportedToken(ERC721UnsupportedToken { token })
+                if token == invalid_operator
+        ));
+    }
+
+    #[motsu::test]
+    fn on_erc721_received_works(
+        contract: Contract<Erc721WrapperTestExample>,
+        erc721_contract: Contract<Erc721>,
+        alice: Address,
+    ) {
+        let token_id = random_token_ids(1)[0];
+
+        contract.init(alice, |contract| {
+            contract.wrapper.underlying.set(erc721_contract.address());
+        });
+
+        let initial_wrapped_balance =
+            contract.sender(alice).erc721.balance_of(alice).motsu_unwrap();
+
+        let operator = erc721_contract.address();
+        let interface_id = contract
+            .sender(operator)
+            .on_erc721_received(operator, alice, token_id, vec![].into())
+            .motsu_expect("should handle ERC721Received");
+
+        assert_eq!(interface_id, RECEIVER_FN_SELECTOR);
+
+        contract.assert_emitted(&erc721::Transfer {
+            from: Address::ZERO,
+            to: alice,
+            token_id,
+        });
+
+        assert_eq!(
+            contract.sender(alice).erc721.balance_of(alice).motsu_unwrap(),
+            initial_wrapped_balance + uint!(1_U256)
+        );
+    }
+
+    #[motsu::test]
+    #[ignore] // TODO: issue in motsu
+    fn recover_reverts_when_invalid_token(
+        contract: Contract<Erc721WrapperTestExample>,
+        erc721_contract: Contract<Erc721>,
+        alice: Address,
+    ) {
+        let token_id = random_token_ids(1)[0];
+        let invalid_token_address = alice;
+
+        contract.init(alice, |contract| {
+            contract.wrapper.underlying.set(invalid_token_address);
+        });
+
+        erc721_contract
+            .sender(alice)
+            ._mint(alice, token_id)
+            .motsu_expect("should mint {token_id} for {alice}");
```

Comment:
```suggestion
    fn recover_reverts_when_invalid_token(
        contract: Contract<Erc721WrapperTestExample>,
        alice: Address,
    ) {
        let token_id = random_token_ids(1)[0];
        let invalid_token_address = alice;

        contract.init(alice, |contract| {
            contract.wrapper.underlying.set(invalid_token_address);
        });
```
there's really no need for `erc721_contract`